### PR TITLE
feat(v2): allow skipping build docs for next version

### LINKED ISSUE #2813: (V2) Option to skip next release docs during build
## 🚀 Feature

In V1, the build command can be given an option so that next release docs are skipped/ignored when the site is built. This helps with the following use case.

Next release docs are not publicly available until actual release . However, we can continue to make content changes to both next release or versioned docs in the meantime.

The build command can accept an option, `skip-next-release`,  and it will skip those contents accordingly and a site be built with only the versioned docs and an appropriate sitemap.

It's been a very useful feature in V1 related to versioning. It'd be great if this can be added to V2.

### Have you read the [Contributing Guidelines on issues](https://github.com/facebook/docusaurus/blob/master/CONTRIBUTING.md#reporting-new-issues)?

Yes

## Motivation

This feature is needed for some of my docs sites that cannot have next release docs available. Having this feature is needed for migrating to V2.

## Pitch

It's already implemented in V1. Theoretically, implementing this in V2 would be simpler since the versioning logic is much simpler than before.

### PR DESCRIPTION
**Pull request by lex111**
## Motivation

Resolve #2813

### Have you read the [Contributing Guidelines on pull requests](https://github.com/facebook/docusaurus/blob/master/CONTRIBUTING.md#pull-requests)?

Yes

## Test Plan

1. In the config file for `docs` set the `excludeNextVersionDocs` field with the `true` value.
1. Try running dev server running `yarn start` and make sure that docs are accessible by `docs/next` path
1. Try building the site running `yarn build` and make sure `build/docs/next` dir does not exist. 
4. Remove `excludeNextVersionDocs` filed and build site again, and make sure `build/docs/next` dir appears. 

## Related PRs

(If this PR adds or changes functionality, please take some time to update the docs at https://github.com/facebook/docusaurus, and link to your PR here.)

### CODE REVIEW

File: website/docs/using-plugins.md [hunk 1]
```diff
@@ -275,6 +275,11 @@ module.exports = {
          * Whether to display the last date the doc was updated.
          */
         showLastUpdateTime: false,
+        /**
+         * Skip the next release docs when versioning is enabled.
+         * This will not build HTML files for documents in `/docs/next` directory, only versioned docs.
+         */
+        skipNextRelease: false,
```
Comment (slorber) [inline website/docs/using-plugins.md:282]: The name seems a bit confusing to me.

What about `excludeNextVersionDocs: true`?

Also worth documenting that this only appliest to build, not both build/start (and maybe the name of the option should reflect that?)

File: packages/docusaurus-plugin-content-docs/src/index.ts [hunk 2]
```diff
@@ -175,25 +176,30 @@ export default function pluginContentDocs(
       // Prepare metadata container.
       const docsMetadataRaw: DocsMetadataRaw = {};
       const docsPromises = [];
+      const includeDefaultDocs = !(
+        options.skipNextRelease && process.argv[2] === 'build'
```
Comment (slorber) [inline packages/docusaurus-plugin-content-docs/src/index.ts:180]: I think we should include the argv[2] as a LoadContext value

Comment (slorber) [reply packages/docusaurus-plugin-content-docs/src/index.ts:180]: also, includeDefaultDocs and skipNextRelease names are not consistent, it's not instantly clear that those 2 names are related to me. 

What about

`const skipNextRelease = options.skipNextRelease && process.argv[2] === 'build';`

(if skipNextRelease is the name we keep)

Comment (lex111) [reply packages/docusaurus-plugin-content-docs/src/index.ts]: I do not see a contradiction in fact, the `includeDefaultDocs` name is pretty clear. This is the final condition, and `skipNextRelease` (`excludeNextVersionDocs`) is just a plugin option.

File: website/docs/using-plugins.md (hunk 1 above)
Comment (lex111) [reply website/docs/using-plugins.md]: Agree, fixed now.

File: packages/docusaurus-plugin-content-docs/src/index.ts (hunk 2 above)
Comment (slorber) [reply packages/docusaurus-plugin-content-docs/src/index.ts]: somehow, it's a name that is the "opposite" of the plugin option.

I'd rather have something like that:

```
const excludeNextVersionDocs = options.excludeNextVersionDocs && process.argv[2] === 'build'

if (!excludeNextVersionDocs) {
  // include next version docs
}
```

probably subjective ;)

**Comment by lex111** [comment]
cc @yangshun

**Comment by slorber** [comment]
lgtm

**Comment by amimas** [comment]
Thanks a lot for bringing this feature from V1. Will this also generate the sitemap without the next release docs if the option is enabled? I couldn't tell from the changes in this PR.

**Comment by slorber** [comment]
yes the next paths should not appear in the sitemaps because we don't add any react router route for the next paths (which are read by the sitemaps plugin)

Tell me if this does not work for you once we release

**Comment by amimas** [comment]
Perfect!! Thank you. Will try it out.

**Pull request merged**